### PR TITLE
[GPU] Fix sub kernel ordering issue in kernels_cache

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *Fix sub kernel ordering code in build_batch in kernels_cache*
 - *...*

### Tickets:
 - *107335*
 - *107542*
 - *107831*

### CODE REVIEW

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 1]
```diff
@@ -562,10 +566,13 @@ void kernels_cache::load(BinaryInputBuffer& ib) {
     }
 }
 
-kernels_cache::compiled_kernels kernels_cache::compile(const kernel_impl_params& params,
+std::vector<kernel::ptr> kernels_cache::compile(const kernel_impl_params& params,
                                             const std::vector<std::shared_ptr<kernel_string>>& kernel_sources,
                                             bool dump_custom_program) {
-    OV_ITT_SCOPED_TASK(ov::intel_gpu::itt::domains::intel_gpu_plugin, "KernelsCache::Compile_ThreadSafe");
+    OV_ITT_SCOPED_TASK(ov::intel_gpu::itt::domains::intel_gpu_plugin, "KernelsCache::compile");
+    if (kernel_sources.empty())
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:573]: When can this happen?

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 2]
```diff
@@ -579,13 +586,23 @@ kernels_cache::compiled_kernels kernels_cache::compile(const kernel_impl_params&
...
 
-    compiled_kernels output_kernels;
+    compiled_kernels _compiled_kernels;
     // Build batches
     for (size_t idx = 0; idx < batches.size(); ++idx) {
-        build_batch(_build_engine, batches[idx], output_kernels);
+        build_batch(_build_engine, batches[idx], _compiled_kernels);
+    }
+
+    OPENVINO_ASSERT(_compiled_kernels.size() == 1, "Only the kernels of the single primitive should be compiled.");
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:595]: Could you rename the function to be more specific, like compile_single_impl ?

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 3]
```diff
@@ -579,13 +586,23 @@ kernels_cache::compiled_kernels kernels_cache::compile(const kernel_impl_params&
...
+
+    auto kernel_data = _compiled_kernels.begin()->second;
+    std::vector<kernel::ptr> output_kernels(kernel_data.size());
+    for (size_t idx = 0; idx < kernel_data.size(); idx++) {
+        size_t part_kernel_idx = kernel_data[idx].second;
+        output_kernels[part_kernel_idx] = kernel_data[idx].first;
     }
 
     t_kernels_code.clear();
+    _compiled_kernels.clear();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:605]: This is temporal variable, do you need this to be cleared manually?

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 4]
```diff
@@ -345,11 +347,13 @@ std::vector<kernel::ptr> kernels_cache::get_kernels(kernel_impl_params params) c
...
+    OPENVINO_ASSERT(res->second.size() != 0, "Number of kernels should not be zero for " + current_node_id);
 
-    std::vector<kernel::ptr> kernels;
-    kernels.reserve(res->second.size());
+    std::vector<kernel::ptr> kernels(res->second.size());
     for (auto& k : res->second) {
-        kernels.emplace_back(k->clone());
+        auto& kernel_ptr = k.first;
+        auto kernel_part_idx = k.second;
+        kernels[kernel_part_idx] = kernel_ptr->clone();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:356]: How about do this in set_kernels ? 
We just can change the parameter type of the set_kernels function.

**Comment by isanghao** [comment]
Performance looks good on dGPU

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp (hunk 4 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:356]: Fixed it

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp (hunk 3 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:573]: Investigate the case kernel_sources is empty and fix more code

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/runtime/kernels_cache.cpp]: Fixed it.

**Review by yeonbok — APPROVED** [review]

**Pull request merged**